### ISSUE TITLE
Empty project built from bacic template loads one core of CPU to 100%

### ISSUE THREAD

**Issue by Sublustris** [defold/defold#9225]
**Describe the bug (REQUIRED)**
Empty project built from bacic template loads one core of CPU to 100% in release and debug versions of application

**To Reproduce (REQUIRED)**
1. Start new project from "desktop" template
2. Click "Project->bundle -> windows application"
3. Run bundled app
4. See  100% load of CPU core

**Defold version (REQUIRED):**
  - Version 1.8.1, 1.9.0, 1.9.1

**Platforms (REQUIRED):**
 - Platforms: Windows
 - OS: Windows 7, 10, 11
 - Device: (tested on several devices)
1)
    Intel(R) Core(TM) i7-10700F CPU @ 2.90GHz (16 CPUs)
    Radeon RX 5700 XT
   16gb ddr4
2)
   Ryzen 7 5800U
   RTX 3070 laptop
   32gb ddr4 

**Minimal repro case project (OPTIONAL):**
standard "Desktop" template

**Screenshots (OPTIONAL):**

![release](https://github.com/user-attachments/assets/11d61776-4a43-4df9-aacf-df650be0e377)
![debug](https://github.com/user-attachments/assets/9278f338-0f10-4054-bfdb-df1ffbb5b0d7)

**Comment by Anisin** [comment]
The problem is not only in an empty project.
On our project MergeVille if I set update frequency = 0 and run via Ctrl + B, my computer starts to get very warm.
Processor 12cpu, 144 hz update frequency

**Comment by Anisin** [comment]
![image](https://github.com/user-attachments/assets/4f440965-8cbe-4d35-9f4d-894373e6ab99)

**Comment by Jhonnyg** [comment]
What happens if you start excluding modules via the appmanifest, do you see a significant difference if you exclude for example the profiler and / or the sound module? The main thread will be synced with the graphics driver if you use VSync, so it must be happening because of "something" going on in one of our support theads 🤔

**Comment by Sublustris** [comment]
> What happens if you start excluding modules via the appmanifest, do you see a significant difference if you exclude for example the profiler and / or the sound module? The main thread will be synced with the graphics driver if you use VSync, so it must be happening because of "something" going on in one of our support theads 🤔

nothing changes.
![manifest](https://github.com/user-attachments/assets/4355c5f3-4936-47f4-ad36-dc9391f0ddcf)
![exclude](https://github.com/user-attachments/assets/6dc64093-a113-4122-a49e-a866515794e0)

**Comment by Jhonnyg** [comment]
And this is with vsync enabled, both in the project and in your OS? i.e not overridden in some GPU vendor control panel or something like that.

**Comment by Sublustris** [comment]
> And this is with vsync enabled, both in the project and in your OS? i.e not overridden in some GPU vendor control panel or something like that.

![изображение](https://github.com/user-attachments/assets/89b25395-dfcb-45f8-bee5-b8485fe7d1f2)

**Comment by Jhonnyg** [comment]
Ok, thanks. And as a last thing, is there any difference if you set "update frequency" to 0? Otherwise I think I'll try to provide you with a windows build that uses the newest GLFW version (platform library for input, window management and so on) and we can maybe see if that helps. I'm not seeing these types of issues on my machine, or perhaps they are not as apparent, so it's a bit hard to debug locally :)

**Comment by Sublustris** [comment]
> Ok, thanks. And as a last thing, is there any difference if you set "update frequency" to 0? Otherwise I think I'll try to provide you with a windows build that uses the newest GLFW version (platform library for input, window management and so on) and we can maybe see if that helps. I'm not seeing these types of issues on my machine, or perhaps they are not as apparent, so it's a bit hard to debug locally :)


There is no difference.

![0](https://github.com/user-attachments/assets/2bc9bc45-614b-48e5-a7bc-d8c83c8fe87a)
![01](https://github.com/user-attachments/assets/76954932-e9dc-425c-a34c-cffded2e1c9f)

**Comment by swajj** [comment]
is this not the same issue as https://github.com/defold/defold/issues/8029 ?

**Comment by Jhonnyg** [comment]
> is this not the same issue as [#8029](https://github.com/defold/defold/issues/8029) ?

No, headless (no graphics adapter) spins as fast as possible afaik, so it will eat 100% cpu. This task is about other threads than the main core that takes 100%. The main thread will be blocked for vsync, so it will not take 100% cpu for an empty project.

**Comment by AGulev** [comment]
I can't reproduce it on computers I have, if you can reproduce it, run build with profiler that shows what different cores do. on Mac it is xcode instruments. Not sure what it is on Windows

**Comment by AGulev** [comment]
I think on windows https://superluminal.eu may help (trial 14 days available)
Could you please build an empty project with app manifest (even empty app manifest), and then take a look what exactly is happening on this busy thread ?

**Comment by AGulev** [comment]
Another option is PIX https://devblogs.microsoft.com/pix/

**Comment by AGulev** [comment]
@Sublustris this is still opened question, and I hope you will provide information I'm asking for

**Comment by AGulev** [comment]
@Sublustris please, if you can reproduce the issue, install https://superluminal.eu/ (it has free trial) and take a look what exactly thread takes CPU time.

**Comment by aglitchman** [comment]
I looked into the issue and, to summarise, GPU waiting is performed by the video driver, which is often a simple spinlock, resulting in 100% load on 1 core. 

The `DwmFlush()` API should more or less solve this problem, which is why Defold switched to the latest version of GLFW some time ago. I have questions:
1) Why is this call disabled for Windows 8 and above in GLFW 3.4 and newer? Or did I misread their code? Look at https://github.com/glfw/glfw/blob/3.4/src/wgl_context.c#L331 (found the commit - https://github.com/glfw/glfw/commit/4005f70eef41a691b849cd06e085d9005df6b66b )
2) It was still enabled in GLFW 3.3 - https://github.com/glfw/glfw/blob/3.3/src/wgl_context.c#L325

Additional  interesting reading (in SDL repo):
* https://github.com/libsdl-org/SDL/issues/5797


P.S. I also got 100% on one core on a computer with the latest drivers and Windows 10 updates. I haven't run any profilers yet.

**Comment by JCash** [comment]
@aglitchman for №1, looking at their git commit it describes the answer a bit: https://github.com/glfw/glfw/commit/4005f70eef41a691b849cd06e085d9005df6b66b

**Comment by AGulev** [comment]
it seems like it should be enabled 

> on Windows 10 or any version of Windows with Desktop Window Manager (DWM) composition 

https://discourse.glfw.org/t/glfwswapinterval/1491

**Comment by AGulev** [comment]
This is strage situation: for some combinations of drivers on new OS in windowed mode this flag should be enabled, but at the same it may me harmful for ppl with another driver+hardware+osVersion combination. As I understand everything I read right.

**Comment by AGulev** [comment]
> [@aglitchman](https://github.com/aglitchman) for №1, looking at their git commit it describes the answer a bit: [glfw/glfw@4005f70](https://github.com/glfw/glfw/commit/4005f70eef41a691b849cd06e085d9005df6b66b)

They made it to fix a fullscreen bug, but as I understand it should be always on in windowed mode all the time.

🤯

**Comment by aglitchman** [comment]
> They made it to fix a fullscreen bug, but as I understand it should be always on in windowed mode all the time.

Yes. It should be like this, judging by the previous commits, comments, and the implementation in other frameworks (SDL, etc.):
1. If app is in full-screen , do nothing, as it has its own synchronisation rules.
2. If app is in windowed and the OS is Windows Vista or newer, check if DWM is enabled and, if so, call DwmFlush().

So the fix is very simple:
`if (!IsWindows8OrGreater() && IsWindowsVistaOrGreater())`
replace with
`if (IsWindowsVistaOrGreater())`

**Comment by aglitchman** [comment]
GLFW 3.4:

![Image](https://github.com/user-attachments/assets/cbe26a65-32ab-4d48-b085-99e7ff06a1fe)

i.e. 1 core pushed to 100%.

GLFW 3.4 + DmwFlush() patch:

![Image](https://github.com/user-attachments/assets/ee29ebae-132d-4101-824a-a007af199913)

**Comment by AGulev** [comment]
Could you pls make a PR for `patch_3.4` file (we have a github action that rebuilds glfw)?

**Comment by aglitchman** [comment]
So, I built GLFW with the fix above for testing, and it works great.

@Sublustris @Anisin Could you please add the dependency to your projects and test?
`https://github.com/aglitchman/ci-test-glfw-34-patch/archive/refs/heads/main.zip`

Add the dependency to the project, click ‘Fetch Libraries’, run the project and check the CPU load.

> Could you pls make a PR for `patch_3.4` file (we have a github action that rebuilds glfw)?

Yes 🙌

**Comment by britzl** [comment]
@aglitchman perhaps also a PR to glfw? If nothing else to get some feedback on why DmwFlush() is not used anymore.

**Comment by AGulev** [comment]
But as I understand it should be ON only in windowed mode, is it takes into account by `DwmIsCompositionEnabled` ?

**Comment by aglitchman** [comment]
> But as I understand it should be ON only in windowed mode, is it takes into account by `DwmIsCompositionEnabled` ?

As far as I understand, this condition checks fullscreen/windowed, so DmwFlush is ON only in windowed mode:
`if (!window->monitor)`

**Comment by aglitchman** [comment]
> If nothing else to get some feedback on why DmwFlush() is not used anymore.

Okay, let's discuss - https://github.com/glfw/glfw/issues/2733

I'll make a PR to Defold a little bit later.

**Comment by aglitchman** [comment]
As a follow-up, I am still looking into and investigating the issue. I received a response in the GLFW issue, but I don't like it.

**Comment by visualmemoryunit** [comment]
Hi. I just wanted to report that I have the same experience on Zorin OS Linux. Running a game (built in defold v. 1.10.4) puts one core to 99-100% regardless if it is windowed, fullscreen, editor or bundled. Just thought I'd mention it because it is a different OS to the original issue.

<img width="574" height="306" alt="Image" src="https://github.com/user-attachments/assets/e46fb4a0-4d3a-40df-bd82-73d0df992baf" />

**Comment by Wolfe2x7** [comment]
Same for Linux Mint 22 Cinnamon, so, some portion of Linux distros.

**Comment by aglitchman** [comment]
_**On Windows and GLFW.**_

After reading a huge amount of opinions and documentation, my conclusions are as follows:
- Calling `DwmFlush()` is a hack that is only needed for bad drivers. It allows you to wait for vsync without a softlock from the video driver. Calling this function has consequences for those who don't have problems: the delta time between frames can fluctuate a lot (0.015-0.017 for 60hz). So this hack isn't an option.
- It seems that the problem only exists for OpenGL. It doesn't exist for DirectX. I can't say for Vulkan, as I don't have any data, but most likely it doesn't exist either (everything is fine on my devices).

I believe that within Windows, the Defold team does not need to do anything except ask users to update their video drivers. In the future, this problem will disappear on its own because the main graphics backends in Defold will be Vulkan and DirectX 12 (correct?), which do not have this problem.

To sum up, this issue can be closed or left open only to figure out the case for Linux 🙌

> [!TIP]
> As a solution for those who have encountered this problem, I can suggest using the ANGLE graphics backend, which uses DirectX 9/11 for graphics output and therefore has no problems with vsync and CPU load: [ANGLE for Defold](https://github.com/indiesoftby/defold-graphics-angle)
>
> (I was inspired by this post https://github.com/defold/defold/issues/11102#issuecomment-3193733459 by @britzl and finally finished the adapter yesterday)

**Comment by britzl** [comment]
> I can suggest using the ANGLE graphics backend, which uses DirectX 9/11 for graphics output and therefore has no problems with vsync and CPU load: [ANGLE for Defold](https://github.com/indiesoftby/defold-graphics-angle)

This is pretty cool! Good job!
